devlib/module/sched.py

```diff
@@ -146,29 +162,56 @@ class SchedDomain(SchedProcFSNode):
-        "SD_SERIALIZE"           : 0x0400,  # Only a single load balancing instance
-        "SD_ASYM_PACKING"        : 0x0800,  # Place busy groups earlier in the domain
-        "SD_PREFER_SIBLING"      : 0x1000,  # Prefer to place tasks in a sibling domain
-        "SD_OVERLAP"             : 0x2000,  # sched_domains of this level overlap
-        "SD_NUMA"                : 0x4000,  # cross-node balancing
+        "SD_LOAD_BALANCE" : SchedDomainFlag(
+            0x0001, "SD_LOAD_BALANCE", "Do load balacing on this domain"),
+        "SD_BALANCE_NEWIDLE" : SchedDomainFlag(
+            0x0002, "SD_BALANCE_NEWIDLE","Balance when about to become idle"),
+        "SD_BALANCE_EXEC" : SchedDomainFlag(
+            0x0004, "SD_BALANCE_EXEC", "Balance on exec"),
+        "SD_BALANCE_FORK" : SchedDomainFlag(
+            0x0008, "SD_BALANCE_FORK", "Balance on fork, clone"),
+        "SD_BALANCE_WAKE" : SchedDomainFlag(
+            0x0010, "SD_BALANCE_WAKE", "Balance on wakeup"),
+        "SD_WAKE_AFFINE" : SchedDomainFlag(
+            0x0020, "SD_WAKE_AFFINE", "Wake task to waking CPU"),
+        "SD_ASYM_CPUCAPACITY" : SchedDomainFlag(
+            0x0040, "SD_ASYM_CPUCAPACITY", "Groups have different max cpu capacities"),
+        "SD_SHARE_CPUCAPACITY" : SchedDomainFlag(
+            0x0080, "SD_SHARE_CPUCAPACITY", "Domain members share cpu capacity"),
+        "SD_SHARE_POWERDOMAIN" : SchedDomainFlag(
+            0x0100, "SD_SHARE_POWERDOMAIN", "Domain members share power domain"),
+        "SD_SHARE_PKG_RESOURCES" : SchedDomainFlag(
+            0x0200, "SD_SHARE_PKG_RESOURCES", "Domain members share cpu pkg resources"),
+        "SD_SERIALIZE" : SchedDomainFlag(
+            0x0400, "SD_SERIALIZE", "Only a single load balancing instance"),
+        "SD_ASYM_PACKING" : SchedDomainFlag(
+            0x0800, "SD_ASYM_PACKING", "Place busy groups earlier in the domain"),
+        "SD_PREFER_SIBLING" : SchedDomainFlag(
+            0x1000, "SD_PREFER_SIBLING", "Prefer to place tasks in a sibling domain"),
+        "SD_OVERLAP" : SchedDomainFlag(
+            0x2000, "SD_OVERLAP", "Sched_domains of this level overlap"),
+        "SD_NUMA" : SchedDomainFlag(
+            0x4000, "SD_NUMA", "Cross-node balancing"),
         # Only defined in Android
         # https://android.googlesource.com/kernel/common/+/android-4.14/include/linux/sched/topology.h#29
-        "SD_SHARE_CAP_STATES"    : 0x8000  # Domain members share capacity state
+        "SD_SHARE_CAP_STATES": SchedDomainFlag(
+            0x8000, "SD_SHARE_CAP_STATES",
+            "(Android only) Domain members share capacity state")
     }
 
     # Checked to be valid from v4.4
     SD_FLAGS_REF_PARTS = (4, 4, 0)
 
+    def __init__(self, nodes):
+        super(SchedDomain, self).__init__(nodes)
+
+        obj_flags = set()
+        for flag in self.SD_FLAGS.values():
+            if self.flags & flag.value == flag.value:
```

Comment:
self.flags does not seem to be defined yet here.

Reply:
It's built automagically from the parent __init__() when parsing the procfs.

Reply:
In that case, it's probably not a good idea to wildly change the type of something created by the mother class itself. There is no way a base class method will be happy with a set of SchedDomainFlag object if it expects an integer. Creating another attribute seems more appropriate, like `flags_obj` maybe ?

Reply:
The way I've wrote up `SchedProcFSNode` is that you still have access to the "raw" read_tree_file procs, but the object view of the procfs is free to use more sophisticated types.

For instance, the flags for a cpu/domain can be obtained via
```python
sd_info.procfs["cpu0"]["domain0"]["flags"] (int)
```
But a much handier version can be obtained via
```python
sd_info.cpus[0].domains[0].flags (set(SchedDomainFlag))
```

So the base classes can depend on the untouched raw procfs version.